src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -8351,111 +8417,1243 @@ class C
         }
 
         [Fact]
-        public void UnmanagedCallersOnlyWithLoopInDefinition_1()
+        public void UnmanagedCallersOnlyCannotDirectCallMethod()
         {
-            var comp = CreateCompilation(@"
-#nullable enable
-namespace System.Runtime.InteropServices
+            var comp = CreateCompilation(new[] { @"
+using System.Runtime.InteropServices;
+public class C
 {
-    [AttributeUsage(AttributeTargets.Method, Inherited = false)]
-    public sealed class UnmanagedCallersOnlyAttribute : Attribute
+    [UnmanagedCallersOnly]
+    public static void M1() { }
+
+    public static void M2()
     {
-        public UnmanagedCallersOnlyAttribute()
+        M1();
+    }
+}
+", UnmanagedCallersOnlyAttribute });
+
+            comp.VerifyDiagnostics(
+                // (10,9): error CS8897: 'C.M1()' is attributed with 'UnmanagedCallersOnly' and cannot be called directly. Obtain a function pointer to this method.
+                //         M1();
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "M1()").WithArguments("C.M1()").WithLocation(10, 9)
+            );
+        }
+
+        [Fact]
+        public void UnmanagedCallersOnlyReferencedFromMetadata()
         {
+            var comp0 = CreateCompilation(new[] { @"
+using System.Runtime.InteropServices;
+public class C
+{
+    [UnmanagedCallersOnly]
+    public static void M1() { }
+}
+", UnmanagedCallersOnlyAttribute });
+
+            validate(comp0.ToMetadataReference());
+            validate(comp0.EmitToImageReference());
+
+            static void validate(MetadataReference reference)
+            {
+                var comp1 = CreateCompilation(@"
+class D
+{
+    public static void M2()
+    {
+        C.M1();
+    }
+}
+", references: new[] { reference });
+
+                comp1.VerifyDiagnostics(
+                    // (6,9): error CS8897: 'C.M1()' is attributed with 'UnmanagedCallersOnly' and cannot be called directly. Obtain a function pointer to this method.
+                    //         C.M1();
+                    Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyMethodsCannotBeCalledDirectly, "C.M1()").WithArguments("C.M1()").WithLocation(6, 9)
+                );
+            }
         }
 
-        public Type[]? CallConvs;
-        public string? EntryPoint;
+        [Fact]
+        public void UnmanagedCallersOnlyReferencedFromMetadata_BadTypeInList()
+        {
+            var il = UnmanagedCallersOnlyAttributeIl + @"
```

Comment:
>var il = UnmanagedCallersOnlyAttributeIl + @" [](start = 12, length = 45)

Might be easier to define in VB with `CreateVisualBasicCompilation()`.

Reply:
Maybe. But I've left a comment as to what the attribute is for this, and there's plenty of other IL-based tests in this file with no VB tests. I'd prefer to leave this as is.

---
In reply to: [483196082](https://github.com/dotnet/roslyn/pull/47311#discussion_r483196082) [](ancestors = 483196082)